### PR TITLE
Add floatingLabelStyle parameter to InputDecoration

### LINKED ISSUE #21385: hintStyle of InputDecorationTheme not correctly applied
Hi all,

When having a custom theme with a custom InputDecorationTheme, like
```dart
new ThemeData(
    //...
    inputDecorationTheme: const InputDecorationTheme(
      focusedBorder: UnderlineInputBorder(borderSide: BorderSide(color: Colors.red)),
      enabledBorder: UnderlineInputBorder(borderSide: BorderSide(color: Colors.yellow)),
      labelStyle: TextStyle(color: Colors.yellow),
      hintStyle: TextStyle(color: Colors.red),
    ));
```
my expectation, according to the documentation of the hintStyle property, is that the label text of a focused text field gets the color of the hint style. Unfortunately it stays the color of the labelStyle. When no custom InputDecorationTheme is in use it works.

------------------------------------
**With** custom theme (expectation: in the focused case the text 'Place' should be red):
custom_theme

------------------------------------

**Without** custom theme (where primary color is cyan and accent color is purple):
no_custom_theme

------------------------------------

Is the configuration wrong or is it really a bug?

Regards,
Florian

### LINKED ISSUE #86389: Add floatingLabelStyle to InputDecoration
Currently our documentation states 

```
  /// When the [labelText] is on top of the input field, the text uses the
  /// [hintStyle] instead.
  final TextStyle? labelStyle;
```

`When the [labelText] is on top of the input field` refers to when the label is sitting in the text field, where typed text would usually be. Our documentation states that in this case the `labelText` should have the style of the `hintStyle`.

This means that when the label is floating (the text field has focus), the `labelText` should use `labelStyle`.

What actually happens is that we use `labelStyle` for when `labelText` in both cases, never using `hintStyle`.

Instead of using `hintStyle` as the style for the label when it is on top of the input field (as stated in the documentation), instead we should have two separate fields `labelStyle` for when the label is sitting on top of the input field, and `floatingLabelStyle` for when the label is floating above the input field. More details on this suggestion: https://github.com/flutter/flutter/issues/21385#issuecomment-419162438 

## Usages
A user may want to set a different style for the `hintText` and the `labelText`. There can only be one inline label on the input field at one time, the `labelText` or the `hintText`, but both can still be defined. In the case where there is no input text, and the text field is focused, then both the `hintText` and `labelText` will be visible. For this case someone may want a `floatingLabelStyle` to pair with the `labelStyle`.

Screen Shot 2021-07-13 at 8 27 41 PM

## Related
#21385

## Proposal
Add a `floatingLabelStyle` property to `InputDecoration` that will be used by the label when it is floating, and correct associated documentation for `labelStyle`. `labelStyle` will be used when the label sits on top of the input field. The `floatingLabelStyle` should default to the `labelStyle` when not defined.

```
InputDecoration(
    labelText: 'label',
    labelStyle: TextStyle(color: Colors.red),
    floatingLabelStyle: TextStyle(color: Colors.green),
)
```

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
## Description
This change adds a `floatingLabelStyle` parameter to `InputDecoration`. When the `labelText` label is floating above the input field, it will use this style. When the `labelText` label is sitting inline with the input field it will use `labelStyle`.  

floatinglabelstyle

## Related Issues
Fixes #86389 
Fixes #21385 

## Tests
I added the following test:
* material/input_decorator_test.dart
  * Widget test to verify input decoration theme overrides are successful. 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 1]
```diff
@@ -2646,16 +2648,23 @@ class InputDecoration {
...
+  /// field, the text uses the [floatingLabelStyle] instead.
   ///
   /// If null, defaults to a value derived from the base [TextStyle] for the
   /// input field and the current [Theme].
   final TextStyle? labelStyle;
 
+  /// The style to use for the [labelText] when the label is above (i.e.,
+  /// vertically adjacent to) the input field.
+  ///
+  /// If null, defaults to a value derived from the base [TextStyle] for the
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/input_decorator.dart:2664]: Shouldn't we specify that [labelStyle] is the default value? That should be required for the sake of backwards compatibility.

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 2]
```diff
@@ -3644,15 +3660,25 @@ class InputDecorationTheme with Diagnosticable {
...
   /// above (i.e., vertically adjacent to) the input field.
   ///
   /// When the [InputDecoration.labelText] is on top of the input field, the
-  /// text uses the [hintStyle] instead.
+  /// text uses the [labelStyle] instead.
   ///
   /// If null, defaults to a value derived from the base [TextStyle] for the
   /// input field and the current [Theme].
-  final TextStyle? labelStyle;
+  final TextStyle? floatingLabelStyle;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/input_decorator.dart:3681]: Same comment as earlier about the default value being the value of [labelStyle].

File: packages/flutter/lib/src/material/input_decorator.dart (hunk 1 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/material/input_decorator.dart]: Correct, fixed.

**Review by HansMuller — APPROVED** [review]
LGTM

**Review by justinmc — APPROVED** [review]
LGTM 👍

**Comment by SimonHausdorf** [comment]
@Renzo-Olivares could it be that `fontSize` isn't accepted due to the animation and scaling of the font?

I tried using `floatingLabelStyle: TextStyle(fontSize: 12),` however, the font appears as like a 9px size instead of a 12.

First shows a 12px font size on a `Text` widget, second a 12px font size on a text input with `floatingLabelStyle`

IMG_35749B039A1F-1

**Pull request merged**